src/main/java/de/berlin/lostberlin/service/fileUpload/FileServiceImpl.java

```diff
@@ -0,0 +1,88 @@
+package de.berlin.lostberlin.service.fileUpload;
+
+import de.berlin.lostberlin.exception.ResourceNotFoundException;
+import de.berlin.lostberlin.exception.ResourceNotSavedException;
+import de.berlin.lostberlin.service.fileUpload.config.FileStorageConfiguration;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.core.io.Resource;
+import org.springframework.core.io.UrlResource;
+import org.springframework.stereotype.Service;
+import org.springframework.util.FileSystemUtils;
+import org.springframework.util.StringUtils;
+import org.springframework.web.multipart.MultipartFile;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.MalformedURLException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.StandardCopyOption;
+
+@Service
+public class FileServiceImpl implements FileService {
+
+    private final Path rootLocation;
+
+    @Autowired
+    public FileServiceImpl(FileStorageConfiguration properties) {
+        this.rootLocation = Paths.get(properties.getLocation());
+    }
+
+    @Override
+    public void initFileStorage() {
+        try {
+            Files.createDirectories(rootLocation);
+        }
+        catch (IOException e) {
+            throw new ResourceNotSavedException("Could not initialize storage");
+        }
+    }
+
+    @Override
+    public Path loadFile(String filename) {
+        return rootLocation.resolve(filename);
+    }
+
+    @Override
+    public void storeFile(MultipartFile file) {
+        String fileName = StringUtils.cleanPath(file.getOriginalFilename());
+
+        try {
+            if (file.isEmpty()) {
+                throw new ResourceNotSavedException("Failed to store empty file " + fileName);
```

Comment:
I would create a separate private method `validate(MultipartFile file)` which implements all these checks and keep storing clean.

It is better to avoid embedded catch/throws, better to keep everything on one level. 

Do you have a special exception handler for `ResourceNotSavedException`?

Reply:
Do you mean smth like this:
 @Override
    public void validateFile(MultipartFile file){
        String fileName = StringUtils.cleanPath(file.getOriginalFilename());
        if (file.isEmpty()) {
            throw new ResourceNotSavedException("Failed to store empty file " + fileName);
        }
        if (fileName.contains("..")) {
            throw new ResourceNotSavedException("Sorry! Filename contains invalid path sequence " + fileName);
        }
    }

    @Override
    public void storeFile(MultipartFile file) {
        validateFile(file);
        String fileName = StringUtils.cleanPath(file.getOriginalFilename());
        try (InputStream inputStream = file.getInputStream()) {
                Files.copy(inputStream, this.rootLocation.resolve(fileName),
                        StandardCopyOption.REPLACE_EXISTING);
        }catch (IOException ex) {
            throw new ResourceNotSavedException("Could not store file " + fileName + ". Please try again!");
        }
    }


There's a handler for ResourceNotSavedException, I've been using this exception for other services as well.